pkg/queue/protobuf_stats_reporter_test.go

```
@@ -0,0 +1,151 @@
+					t.Errorf("handler returned wrong status code: got %v want %v",
+						status, http.StatusInternalServerError)
+				}
+				if rr.Body.String() != expected {
+					t.Errorf("handler returned unexpected body: got %v want %v",
+						rr.Body.String(), expected)
+				}
+			} else { // good case, data received
+				bodyBytes, err := ioutil.ReadAll(rr.Body)
+				if err != nil {
+					t.Errorf("reading body failed: %w", err)
+				}
+				stat := metrics.Stat{}
+				err = stat.Unmarshal(bodyBytes)
+				if err != nil {
+					t.Errorf("unmarshalling failed: %w", err)
+				}
+				if stat.PodName != testStat.PodName {
+					t.Errorf("handler returned wrong stat data: got stat.PodName = %v want %v",
+						stat.PodName, testStat.PodName)
+				}
+				if stat.RequestCount != testStat.RequestCount {
+					t.Errorf("handler returned wrong stat data: got stat.RequestCount = %v want %v",
+						stat.RequestCount, testStat.RequestCount)
+				}
+				if stat.ProcessUptime != testStat.ProcessUptime {
+					t.Errorf("handler returned wrong stat data: got stat.ProcessUptime = %v want %v",
+						stat.ProcessUptime, testStat.ProcessUptime)
+				}
+				if stat.ProxiedRequestCount != testStat.ProxiedRequestCount {
+					t.Errorf("handler returned wrong stat data: got stat.ProxiedRequestCount = %v want %v",
+						stat.ProxiedRequestCount, testStat.ProxiedRequestCount)
+				}
+				if stat.AverageProxiedConcurrentRequests != testStat.AverageProxiedConcurrentRequests {
+					t.Errorf("handler returned wrong stat data: got AverageProxiedConcurrentRequests = %v want %v",
+						stat.AverageProxiedConcurrentRequests, testStat.AverageProxiedConcurrentRequests)
+				}
+				if stat.AverageConcurrentRequests != testStat.AverageConcurrentRequests {
+					t.Errorf("handler returned wrong stat data: got stat.AverageConcurrentRequests = %v want %v",
+						stat.AverageConcurrentRequests, testStat.AverageConcurrentRequests)
+				}
+			}
+		})
+	}
```

Btw this part and any other part that iterates on structs could be compacted as follows but with very limited readability:
```
val1 := reflect.ValueOf(&stat).Elem()
val2 := reflect.ValueOf(&testStat).Elem()

for i := 0; i < val1.NumField(); i = i + 1 {
	val1.Field(i).Interface()
	if val1.Field(i).Interface() != val2.Field(i).Interface() {
		t.Errorf("handler returned wrong stat data: got %s = %v want %v", val1.Type().Field(i).Name,
		val1.Field(i).Interface(), val2.Field(i).Interface())
```

---

Since I ditched `time.Time` off the metric, we can probably switch to using `cmp.Equal` and `cmp.Diff` here. They should do roughly what you've written out there.

---

Yeah sure I guess they fail fast and report per field name.